feat: Add the add-and-remove task feature

gvgesanto2 opened the pull request:
## Description

**Task:** https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md

This PR makes the following changes:
-  Adds the ``add task`` and ``remove task`` features
- Adds the ``set task to completed`` and ``edit task description`` features
- Stores all the changes made in the tasks list into the browser's local storage

Plus, it adopts the following methodologies and conventions:
- For the CSS/Sass:
  - [BEM](https://en.bem.info/methodology/) methodology for naming the CSS classes
  - A modified version of the [Scalable and Modular Architecture for CSS](http://smacss.com/) (SMACSS) methodology for organizing the CSS code into categorized folders/files (base, components, layout, and utilities)
  - [Atomic Design](https://bradfrost.com/blog/post/atomic-web-design/) methodology for breaking down the CSS components into atoms, molecules, and organisms
- For the JavaScript:
  - [MVC architecture](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) to separate the application business logic from the presentation logic (DOM manipulation tasks)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [X] My code follows the style guidelines of this project (no linters errors)
- [X] I have checked my code and corrected any misspellings
- [X] I have commented my code
- [X] I have deleted any unused code as comments in my codebase

## Screenshots

![todo-add-remove](https://user-images.githubusercontent.com/64566209/171913308-7d57e49f-37c4-45a8-beb5-84387a8c9140.png)

gvgesanto2 committed:
feat: Add the adding new task feature

gvgesanto2 committed:
feat: Add the set completed and edit description features

gvgesanto2 committed:
feat: Add the remove task feature

gvgesanto2 committed:
fix: Fix linters errors

KDlamini commented on src/scripts/controllers/TasksListController.js:
```diff
@@ -1,14 +1,57 @@
+    }
+  }
+
+  handleRemoveTask = (taskId) => {
+    this.tasksList.removeTaskById(taskId);
+    this.tasksListView.removeTaskFromScreen(taskId);
+    this.updateStorage();
   }
 }
```
Deleting a task should update all remaining items' indexes, so they represent the current list order and are unique. Currently, your indexes are not updating when an item is deleted.

- in the example below I added 6 items, then I deleted the 2nd and 5th items and this is how your indexes look like:

![image](https://user-images.githubusercontent.com/38088818/171918023-ab204c34-2ae9-49da-b805-c94726f14027.png)

gvgesanto2 replied:
## Hi @KDlamini 👋,

**Thank you so much for the feedback!** About the required change of updating the tasks ``indexes`` when the user deletes a task, I'm aware this was one of the requirements, but sometimes as a software engineer is our job to try to understand the ``why`` instead of just following the requirements blindly. 

The way I see it, the reason for updating the tasks indexes is:
- to always render the task list in order
- to have always a unique ``index/id`` for each task

**My solution does that!** Even when adding new tasks or deleting an existing one, the tasks ``ids/indexes`` represents the current list order (there are 'gaps' between the id's numbers but they are always in order, always in sequence) and are unique for each task. 

The reason I did this way instead of the way it was suggested in the requirements is that **iterating through an array can be a VERY expensive operation and we should, in my opinion, avoid doing that if it's unnecessary.** 
In this case, for example, I believe it's totally unnecessary. This won't give me any new functionality, this won't improve my code quality, and this won't give me ANY benefit over my solution at all.  To be honest, this will actually make my code less efficient than it is right now... 
Imagine if a had a tasks list with ``100,000 tasks`` in it. In this case, when the user deletes the first task, for example, I would have to update the index of ``99,999 tasks``. That's a very expensive and inefficient operation. This is the worst-case scenario, I know, but in the average case scenario where the user deletes ``the task in the middle with the index of #50,000``, I would still have to update ``49,999 tasks``! And the worst thing is I would be iterating through this huge array EVERY TIME the user deletes a task!

**Note:** my solution for the tasks ``ids/indexes`` also allows me to implement the dragging feature with the same efficiency that if I had implemented the solution it was suggested in the requirements.

Being that said, I can refactor my code to use the requirement's solution, but **I strongly doubt this is the best way to address this problem** (I'm not saying my solution is the best approach either, but it is definitely better than the one it was suggested in the requirements).

KDlamini reviewed (changes requested):
### Hi @gvgesanto2 🙋‍♀️,  

Great job so far! 💯 
There are some issues that you still need to work on improving your project according to the requirements. You are almost there! 💪🏾 

![pugReaching](https://user-images.githubusercontent.com/38088818/171918694-05eba4ea-6a62-4703-90d9-7689e77c8616.gif)

## Highlights 🌞 
✔️  Good use of the MVC pattern, however, be careful not to over-complicate your code.
✔️  Good file structure.
✔️  Great PR title and description
✔️  Correct use of Gitflow 
✔️  Commit messages are meaningful
✔️  No linter errors

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
N/A

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Lucas-Masaba reviewed (changes requested):
# Hello @gvgesanto2  👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
![almost-there](https://media.giphy.com/media/dEPGknRiEniec/giphy.gif)


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions 💭 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏 💻 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

- [x] Helo @gvgesanto2 👋 . I read your suggestion you wrote earlier to the previous pull request. I talked it over with a senior and we came to a conclusion. I agree with your argument but you are required to follow the instructions given to you for the milestone. You can write about your suggestion to [student success](https://microverse.zendesk.com/hc/en-us/requests/new) but you are still required to make sure that your indices are updating.

gvgesanto2 committed:
feat: Update the tasks indexes when the user deletes a task

gvgesanto2 commented:
## Description

**Task:** https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md

This PR addresses the following change requested by the last and [first review](https://github.com/gvgesanto2/microverse-todo-list/pull/2#discussion_r889196827):

> Deleting a task should update all remaining items' indexes, so they represent the current list order and are unique. Currently, your indexes are not updating when an item is deleted.

- The tasks' indexes are updated every time the user deletes a task, representing the current list order, the exact same way as it was suggested in this project's requirements.

## Checklist:

- [X] My code follows the style guidelines of this project (no linters errors)
- [X] I have checked my code and corrected any misspellings
- [X] I have commented my code
- [X] I have deleted any unused code as comments in my codebase

## Screenshots

![todo-index](https://user-images.githubusercontent.com/64566209/171957418-bd9dde39-5abb-45f5-a7c3-fba2f1accf09.png)

OmarRamoun reviewed (approved):
# 🟢 Project Approved! 🏆

## Hi @gvgesanto2 👋

Your project is complete! ✔️ 🏁

Finally, **you did it**! 🎉 👏. Your code was very professional 💯 and I can see you have a great chance to be a Great Software Engineer ⛑️.

There is nothing else to say other than... _it's time to merge it_ :shipit:
Congratulations! 🎉. And Welcome to Microverse 👋.

> **Note:** You can now take a break even if short before you move on to the next project. Happy 😄 Coding 🖥️! and Have a nice day 🌞.

## Highlights 🌟

- ✔️ Descriptive commmit messsages
- ✔️ Descriptive PR description
- ✔️ No linter Errors
- The code is top-notch! 💯 and have a huge potential.

Cheers and Happy coding!👏👏👏

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [![@omarramoun](https://img.shields.io/github/followers/omarramoun?label=Ramoun&style=social)](https://github.com/omarramoun)
- This is My **Twitter**: [![@omarramoun](https://img.shields.io/twitter/follow/omarramoun?label=ramoun16&style=social)](https://twitter.com/omarramoun)
- This is My **Slack**: [![Ramoun](https://img.shields.io/badge/slack-@Ramoun-brightgreen)](https://slack.com)
------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
